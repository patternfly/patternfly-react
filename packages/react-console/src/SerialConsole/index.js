export { default as SerialConsole } from './SerialConsole';
