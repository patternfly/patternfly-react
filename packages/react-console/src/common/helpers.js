import React from 'react';

/** Implementation of the debounce function */
export const debounce = (func, wait) => {
  let timeout;
  function innerFunc(...args) {
    const context = this;
    clearTimeout(timeout);
    timeout = setTimeout(() => func.apply(context, args), wait);
  }
  return innerFunc;
};

export const child = (children, index) => {
  const childrenArray = React.Children.toArray(children);
  return (childrenArray && childrenArray.length > index && childrenArray[index]) || null;
};
