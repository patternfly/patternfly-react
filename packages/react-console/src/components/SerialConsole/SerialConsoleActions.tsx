import React from 'react';
import { css } from '@patternfly/react-styles';
import { Button } from '@patternfly/react-core';

import styles from '@patternfly/react-styles/css/components/Consoles/SerialConsole';

export interface SerialConsoleActionsProps extends React.HTMLProps<HTMLDivElement> {
  onDisconnect: () => void;
  onReset: () => void;
  textDisconnect?: string;
  textReset?: string;
}

export const SerialConsoleActions: React.FunctionComponent<SerialConsoleActionsProps> = ({
  textDisconnect = 'Disconnect',
  textReset = 'Reset',
  ...props
}: SerialConsoleActionsProps) => (
  <div className={css(styles.consoleActionsSerial)}>
    <Button variant="secondary" onClick={props.onDisconnect}>
      {textDisconnect}
    </Button>
    <Button variant="secondary" onClick={props.onReset}>
      {textReset}
    </Button>
  </div>
);
SerialConsoleActions.displayName = 'SerialConsoleActions';
