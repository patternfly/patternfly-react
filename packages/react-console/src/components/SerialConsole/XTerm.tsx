import React, { useImperativeHandle } from 'react';

import { Terminal } from 'xterm';
import { FitAddon } from 'xterm-addon-fit';

import { debounce, canUseDOM } from '@patternfly/react-core';

export interface XTermProps {
  /** The number of columns to resize to */
  cols?: number;
  /** The number of rows to resize to */
  rows?: number;
  fontFamily?: string;
  fontSize?: number;
  /** Terminal title has been changed. */
  onTitleChanged?: (title: string) => void;
  /** Data to be sent from terminal to backend; (data) => {} */
  onData?: (e: string) => void;
  /** A reference object to attach to the xterm */
  innerRef?: React.RefObject<any>;
}

export const XTerm: React.FunctionComponent<XTermProps> = ({
  cols = 80,
  rows = 25,
  fontFamily,
  fontSize,
  onTitleChanged,
  onData,
  innerRef
}) => {
  let terminal: Terminal;
  const ref = React.useRef<HTMLDivElement>();

  useImperativeHandle(innerRef, () => ({
    focusTerminal() {
      if (terminal) {
        terminal.focus();
      }
    },
    /**
     * Backend sent data.
     *
     * @param {string} data String content to be writen into the terminal
     */
    onDataReceived: (data: string) => {
      if (terminal) {
        terminal.write(data);
      }
    },
    /**
     * Backend closed connection.
     *
     * @param {string} reason String error to be written into the terminal
     */
    onConnectionClosed: (reason: string) => {
      if (terminal) {
        terminal.write(`\x1b[31m${reason || 'disconnected'}\x1b[m\r\n`);
        terminal.refresh(terminal.rows, terminal.rows); // start to end row
      }
    }
  }));

  React.useEffect(() => {
    const fitAddon = new FitAddon();
    terminal = new Terminal({
      cols,
      rows,
      cursorBlink: true,
      fontFamily,
      fontSize,
      screenReaderMode: true
    });

    const onWindowResize = () => {
      const geometry = fitAddon.proposeDimensions();
      if (geometry) {
        terminal.resize(geometry.rows, geometry.cols);
      }
    };

    if (onData) {
      terminal.onData(onData);
    }

    if (onTitleChanged) {
      terminal.onTitleChange(onTitleChanged);
    }

    terminal.loadAddon(fitAddon);

    terminal.open(ref.current);

    const resizeListener = debounce(onWindowResize, 100);

    if (!rows) {
      if (canUseDOM) {
        window.addEventListener('resize', resizeListener);
      }
      onWindowResize();
    }
    terminal.focus();

    return () => {
      terminal.dispose();
      if (canUseDOM) {
        window.removeEventListener('resize', resizeListener);
      }
      onFocusOut();
    };
  }, []);

  const onBeforeUnload = (event: any) => {
    // Firefox requires this when the page is in an iframe
    event.preventDefault();

    // see "an almost cross-browser solution" at
    // https://developer.mozilla.org/en-US/docs/Web/Events/beforeunload
    event.returnValue = '';
    return '';
  };

  const onFocusIn = () => {
    window.addEventListener('beforeunload', onBeforeUnload);
  };

  const onFocusOut = () => {
    window.removeEventListener('beforeunload', onBeforeUnload);
  };

  // ensure react never reuses this div by keying it with the terminal widget
  // Workaround for xtermjs/xterm.js#3172
  return <div ref={ref} className="pf-c-console__xterm" role="list" onFocus={onFocusIn} onBlur={onFocusOut} />;
};
XTerm.displayName = 'XTerm';
