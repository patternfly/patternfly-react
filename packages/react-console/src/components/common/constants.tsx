const NONE_TYPE = '_none_';
const SERIAL_CONSOLE_TYPE = 'SerialConsole';
const SPICE_CONSOLE_TYPE = 'SpiceConsole';
const VNC_CONSOLE_TYPE = 'VncConsole';
const RDP_CONSOLE_TYPE = 'RdpConsole';
const DESKTOP_VIEWER_CONSOLE_TYPE = 'DesktopViewer';

const CONNECTING = 'connecting';
const CONNECTED = 'connected';
const DISCONNECTED = 'disconnected';
const LOADING = 'loading';

const DEFAULT_VV_FILENAME = 'console.vv';
const DEFAULT_VV_MIMETYPE = 'application/x-virt-viewer';
const DEFAULT_RDP_FILENAME = 'console.rdp';
const DEFAULT_RDP_MIMETYPE = 'application/rdp';
const DEFAULT_RDP_PORT = 3389;

export const constants = {
  NONE_TYPE,
  SERIAL_CONSOLE_TYPE,
  SPICE_CONSOLE_TYPE,
  VNC_CONSOLE_TYPE,
  RDP_CONSOLE_TYPE,
  DESKTOP_VIEWER_CONSOLE_TYPE,

  CONNECTING,
  CONNECTED,
  DISCONNECTED,
  LOADING,

  DEFAULT_VV_FILENAME,
  DEFAULT_VV_MIMETYPE,
  DEFAULT_RDP_FILENAME,
  DEFAULT_RDP_MIMETYPE,
  DEFAULT_RDP_PORT
};
