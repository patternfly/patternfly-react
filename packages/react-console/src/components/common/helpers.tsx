import React from 'react';

export const child = (children: React.ReactNode, index: number) => {
  const childrenArray = React.Children.toArray(children);
  return (childrenArray && childrenArray.length > index && childrenArray[index]) || null;
};
