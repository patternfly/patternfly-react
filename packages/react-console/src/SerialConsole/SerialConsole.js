import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { EmptyState, Button, noop } from 'patternfly-react';
import { CONNECTED, DISCONNECTED, LOADING } from './constants';

import XTerm from './XTerm';
import SerialConsoleActions from './SerialConsoleActions';

/**
 * SerialConsole Component for PatternFly React
 */
class SerialConsole extends React.Component {
  componentDidMount() {
    this.props.onConnect();
  }

  componentWillUnmount() {
    this.props.onDisconnect();
  }

  onResetClick = event => {
    if (event.button !== 0) return;

    this.props.onDisconnect();
    this.props.onConnect();
    event.target.blur();
    this.focusTerminal();
  };

  onDisconnectClick = event => {
    if (event.button !== 0) return;

    this.props.onDisconnect();
    event.target.blur();
    this.focusTerminal();
  };

  /**
   * Backend sent data.
   */
  onDataReceived(data) {
    if (this.childTerminal && this.props.status === CONNECTED) {
      this.childTerminal.onDataReceived(data);
    }
  }

  /**
   * Backend closed connection.
   */
  onConnectionClosed(reason) {
    if (this.childTerminal) {
      this.childTerminal.onConnectionClosed(reason);
    }
  }

  focusTerminal = () => {
    this.childTerminal && this.childTerminal.focus();
  };

  render() {
    const { id, status, topClassName } = this.props;
    const idPrefix = `${id || 'id'}-serialconsole`;

    let terminal;
    let isDisconnectEnabled = false;
    switch (status) {
      case CONNECTED:
        terminal = (
          <XTerm
            ref={c => {
              this.childTerminal = c;
            }}
            className="serial-console-pf"
            cols={this.props.cols}
            rows={this.props.rows}
            autoFit={this.props.autoFit}
            fontFamily={this.props.fontFamily}
            fontSize={this.props.fontSize}
            onConnect={this.props.onConnect}
            onDisconnect={this.props.onDisconnect}
            onData={this.props.onData}
            onTitleChanged={this.props.onTitleChanged}
            onResize={this.props.onResize}
          />
        );
        isDisconnectEnabled = true;
        break;
      case DISCONNECTED:
        terminal = (
          <EmptyState>
            <EmptyState.Title>{this.props.textDisconnectedTitle}</EmptyState.Title>
            <EmptyState.Info>{this.props.textDisconnected}</EmptyState.Info>
            <EmptyState.Action>
              <Button bsStyle="primary" bsSize="large" onClick={this.props.onConnect}>
                {this.props.textConnect}
              </Button>
            </EmptyState.Action>
          </EmptyState>
        );
        break;
      case LOADING:
      default:
        terminal = <span>{this.props.textLoading}</span>;
        break;
    }

    const classes = classNames('serial-console-pf', topClassName);

    return (
      <div className={classes} id={id}>
        <SerialConsoleActions
          idPrefix={idPrefix}
          isDisconnectEnabled={isDisconnectEnabled}
          onDisconnect={this.onDisconnectClick}
          onReset={this.onResetClick}
          textDisconnect={this.props.textDisconnect}
          textReconnect={this.props.textReconnect}
        />
        <div className="panel-body console-terminal-pf">{terminal}</div>
      </div>
    );
  }
}

SerialConsole.propTypes = {
  /** Initiate connection to backend. In other words, the calling components manages connection state. */
  onConnect: PropTypes.func.isRequired,
  /** Close connection to backend */
  onDisconnect: PropTypes.func.isRequired,
  /** Terminal has been resized, backend shall be informed. (rows, cols) => {} */
  onResize: PropTypes.func,
  /** Terminal produced data, like key-press */
  onData: PropTypes.func,
  /** Terminal title has been changed. */
  onTitleChanged: PropTypes.func,

  /** Connection status, a value from [''connected', 'disconnected', 'loading']. Default is 'loading' for a not matching value. */
  status: PropTypes.string.isRequired,
  id: PropTypes.string,

  /** Size of the terminal component */
  rows: PropTypes.number,
  cols: PropTypes.number,
  autoFit: PropTypes.bool,

  /** Font for text rendered to xterm canvas */
  fontFamily: PropTypes.string,
  fontSize: PropTypes.number,

  /** Enable customization */
  topClassName: PropTypes.string,

  /** Localization */
  textDisconnect: PropTypes.string,
  textDisconnectedTitle: PropTypes.string,
  textDisconnected: PropTypes.string,
  textLoading: PropTypes.string,
  textReconnect: PropTypes.string,
  textConnect: PropTypes.string
};

SerialConsole.defaultProps = {
  topClassName: '',

  id: '',
  rows: 25,
  cols: 80,
  autoFit: false,

  fontFamily: undefined /** Use xterm default: 'courier-new, courier, monospace' */,
  fontSize: undefined /** Use xterm default: 15 */,

  onTitleChanged: noop,
  onData: noop,
  onResize: noop,

  textDisconnectedTitle: 'Disconnected from serial console',
  textDisconnected: 'Click Connect to open serial console.',
  textLoading: 'Loading ...',
  textConnect: 'Connect',
  textDisconnect: undefined /** Default is set in SerialConsoleActions */,
  textReconnect: undefined /** Default is set in SerialConsoleActions */
};

export default SerialConsole;
