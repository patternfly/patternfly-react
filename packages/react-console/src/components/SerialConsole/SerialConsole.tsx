import React from 'react';

import { css } from '@patternfly/react-styles';
import { Button, EmptyState, EmptyStateBody, EmptyStateIcon, Spinner } from '@patternfly/react-core';

import { XTerm, XTermProps } from './XTerm';
import { SerialConsoleActions } from './SerialConsoleActions';

import { constants } from '../common/constants';

import styles from '@patternfly/react-styles/css/components/Consoles/SerialConsole';
import '@patternfly/react-styles/css/components/Consoles/xterm.css';
import '@patternfly/react-styles/css/components/Consoles/SerialConsole.css';

const { CONNECTED, DISCONNECTED, LOADING } = constants;

export interface SerialConsoleProps extends XTermProps {
  /** Initiate connection to backend. In other words, the calling components manages connection state. */
  onConnect: () => void;
  /** Close connection to backend */
  onDisconnect: () => void;
  /** Terminal produced data, like key-press */
  onData: (e: string) => void;
  /** Terminal title has been changed */
  onTitleChanged?: () => void;
  /** Connection status; a value from [''connected'; 'disconnected'; 'loading']. Default is 'loading' for a not matching value. */
  /** The number of columns to resize to */
  cols?: number;
  /** The number of rows to resize to */
  rows?: number;
  fontFamily?: string;
  fontSize?: number;
  status?: string;
  /** Text content rendered inside the Connect button */
  textConnect?: string;
  /** Text content rendered inside the Disconnect button */
  textDisconnect?: string;
  /** Text content rendered inside the Reset button */
  textReset?: string;
  /* Text content rendered inside the EmptyState for when console is disconnnected */
  textDisconnected?: string;
  /* Text content rendered inside the EmptyState for when console is loading */
  textLoading?: string;
  /** A reference object to attach to the SerialConsole. */
  innerRef?: React.RefObject<any>;
}

const SerialConsoleBase: React.FunctionComponent<SerialConsoleProps> = ({
  onConnect,
  onDisconnect,
  onTitleChanged = () => {},
  onData,
  cols,
  rows,
  fontFamily,
  fontSize,
  status = 'loading',
  textConnect = 'Connect',
  textDisconnect,
  textReset,
  textDisconnected = 'Click Connect to open serial console.',
  textLoading = 'Loading ...',
  innerRef
}) => {
  React.useEffect(() => {
    onConnect();
    return () => {
      onDisconnect();
    };
  }, []);

  const onConnectClick = () => {
    onConnect();
    focusTerminal();
  };

  const onDisconnectClick = () => {
    onDisconnect();
    focusTerminal();
  };

  const onResetClick = () => {
    onDisconnect();
    onConnect();
    focusTerminal();
  };

  const focusTerminal = () => {
    innerRef && innerRef.current && innerRef.current.focusTerminal();
  };

  let terminal;
  switch (status) {
    case CONNECTED:
      terminal = (
        <XTerm
          innerRef={innerRef}
          cols={cols}
          rows={rows}
          fontFamily={fontFamily}
          fontSize={fontSize}
          onTitleChanged={onTitleChanged}
          onData={onData}
        />
      );
      break;
    case DISCONNECTED:
      terminal = (
        <EmptyState>
          <EmptyStateBody>{textDisconnected}</EmptyStateBody>
          <Button onClick={onConnectClick}>{textConnect}</Button>
        </EmptyState>
      );
      break;
    case LOADING:
    default:
      terminal = (
        <EmptyState>
          <EmptyStateIcon variant="container" component={Spinner} />
          <EmptyStateBody>{textLoading}</EmptyStateBody>
        </EmptyState>
      );
      break;
  }

  return (
    <>
      {status !== DISCONNECTED && (
        <SerialConsoleActions
          onDisconnect={onDisconnectClick}
          onReset={onResetClick}
          textDisconnect={textDisconnect}
          textReset={textReset}
        />
      )}
      <div className={css(styles.consoleSerial)}>{terminal}</div>
    </>
  );
};
SerialConsoleBase.displayName = 'SerialConsoleBase';
export const SerialConsole = React.forwardRef((props: SerialConsoleProps, ref: React.Ref<HTMLDivElement>) => (
  <SerialConsoleBase innerRef={ref as React.MutableRefObject<any>} {...props} />
));
SerialConsole.displayName = 'SerialConsole';
