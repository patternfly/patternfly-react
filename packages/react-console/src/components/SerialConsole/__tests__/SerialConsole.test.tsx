import React from 'react';

import { render } from '@testing-library/react';

import { SerialConsole } from '../SerialConsole';
import { constants } from '../../common/constants';

const { CONNECTED, DISCONNECTED, LOADING } = constants;

describe('SerialConsole', () => {
  beforeAll(() => {
    window.HTMLCanvasElement.prototype.getContext = () => ({ canvas: {} } as any);
  });

  test('in the LOADING state', () => {
    const { asFragment } = render(
      <SerialConsole
        onData={jest.fn()}
        onConnect={jest.fn()}
        onDisconnect={jest.fn()}
        status={LOADING}
        textLoading="My text for Loading"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('in the DISCONNECTED state', () => {
    const { asFragment } = render(
      <SerialConsole
        onData={jest.fn()}
        onConnect={jest.fn()}
        onDisconnect={jest.fn()}
        status={DISCONNECTED}
        textDisconnected="My text for Disconnected"
        textConnect="My text for Connect"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('with CONNECTED state', () => {
    beforeAll(() => {
      window.HTMLCanvasElement.prototype.getContext = () =>
        ({ canvas: {}, createLinearGradient: jest.fn(), fillRect: jest.fn() } as any);
      global.window.matchMedia = () => ({
        addListener: jest.fn(),
        removeListener: jest.fn(),
        matches: true,
        media: undefined,
        onchange: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn()
      });
    });

    test('renders', () => {
      const { asFragment } = render(
        <SerialConsole
          cols={33}
          rows={44}
          status={CONNECTED}
          onData={jest.fn()}
          onConnect={jest.fn()}
          onDisconnect={jest.fn()}
          textDisconnect="My text for Disconnect"
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
