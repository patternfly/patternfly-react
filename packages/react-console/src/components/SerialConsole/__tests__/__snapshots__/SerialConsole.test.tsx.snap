// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render SerialConsole in the CONNECTED state 1`] = `
Array [
  <div
    class="pf-c-console__actions-serial"
  >
    <button
      aria-disabled="false"
      class="pf-c-button pf-m-secondary"
      data-ouia-component-id="OUIA-Generated-Button-secondary-1"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe="true"
      type="button"
    >
      My text for Disconnect
    </button>
    <button
      aria-disabled="false"
      class="pf-c-button pf-m-secondary"
      data-ouia-component-id="OUIA-Generated-Button-secondary-2"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe="true"
      type="button"
    >
      Reset
    </button>
  </div>,
  <div
    class="pf-c-console__serial"
  >
    <div
      class="pf-c-console__xterm"
      role="list"
    />
  </div>,
]
`;

exports[`SerialConsole in the CONNECTED state 1`] = `
<SerialConsoleBase
  cols={33}
  id="myidprefix"
  innerRef={null}
  onConnect={[MockFunction]}
  onDisconnect={[MockFunction]}
  rows={44}
  status="connected"
  textDisconnect="My text for Disconnect"
  textReconnect="My text for Reconnect"
/>
`;

exports[`SerialConsole in the DISCONNECTED state 1`] = `
<SerialConsoleBase
  id="myidprefix"
  innerRef={null}
  onConnect={[MockFunction]}
  onDisconnect={[MockFunction]}
  status="disconnected"
  textConnect="My text for Connect"
  textDisconnected="My text for Disconnected"
  textDisconnectedTitle="My title for Disconnected"
/>
`;

exports[`SerialConsole in the LOADING state 1`] = `
<SerialConsoleBase
  innerRef={null}
  onConnect={[MockFunction]}
  onDisconnect={[MockFunction]}
  status="loading"
  textLoading="My text for Loading"
/>
`;
